File: scripts/requeue_needs_retry.py
"""Re-queue Needs Retry jobs (Python fallback when tsx/better-sqlite3 unavailable)."""
import os
import re
import sqlite3

PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DB_PATH = os.path.join(PROJECT_ROOT, "jobagent.sqlite")
JOBS_DIR = os.path.join(PROJECT_ROOT, "jobs")
SUBMISSIONS_DIR = os.path.join(PROJECT_ROOT, "submissions")
MAX_AUTO_RETRIES = 3


def resolve_folder(company: str) -> str | None:
    slug = re.sub(r"[^a-z0-9]+", "_", company.lower()).strip("_")
    direct = os.path.join(SUBMISSIONS_DIR, slug)
    if os.path.isdir(direct):
        return direct
    if not os.path.isdir(SUBMISSIONS_DIR):
        return None
    stripped = re.sub(r"[^a-z0-9]", "", company.lower())
    for name in os.listdir(SUBMISSIONS_DIR):
        full = os.path.join(SUBMISSIONS_DIR, name)
        if os.path.isdir(full) and re.sub(r"[^a-z0-9]", "", name.lower()) == stripped:
            return full
    return None


def main():
    conn = sqlite3.connect(DB_PATH)
    jobs = conn.execute(
        "SELECT id, company, title, url, COALESCE(retry_count, 0) FROM jobs WHERE status = 'Needs Retry'"
    ).fetchall()

    if not jobs:
        print("No jobs marked Needs Retry.")
        return

    os.makedirs(JOBS_DIR, exist_ok=True)
    to_drafted = to_new = skipped = 0

    for job_id, company, title, url, retry_count in jobs:
        if retry_count >= MAX_AUTO_RETRIES:
            print(f"  -> Skip {company} — max auto-retries ({MAX_AUTO_RETRIES}) reached.")
            skipped += 1
            continue

        folder = resolve_folder(company)
        original_jd = os.path.join(folder, "Original_JD.txt") if folder else None
        slug = re.sub(r"[^a-z0-9]+", "_", company, flags=re.I).strip("_")
        staging = os.path.join(JOBS_DIR, f"{slug}_{job_id[:8]}.txt")

        if original_jd and os.path.isfile(original_jd):
            content = open(original_jd, encoding="utf-8-sig").read()
            if url and not content.strip().startswith("URL:"):
                content = f"URL: {url}\n\n{content}"
            with open(staging, "w", encoding="utf-8") as f:
                f.write(content)
            conn.execute("UPDATE jobs SET status = 'Drafted' WHERE id = ?", (job_id,))
            print(f"  -> {company}: restored JD -> Drafted ({os.path.basename(staging)})")
            to_drafted += 1
        elif url and not str(url).startswith("local://"):
            conn.execute("UPDATE jobs SET status = 'New' WHERE id = ?", (job_id,))
            print(f"  -> {company}: will re-scrape -> New")
            to_new += 1
        else:
            print(f"  -> Skip {company} — no saved JD or URL to requeue.")
            skipped += 1

    conn.commit()
    conn.close()
    print(f"[REQUEUE_DONE] {to_drafted} -> Drafted, {to_new} -> New, {skipped} skipped.")


if __name__ == "__main__":
    main()

File: scripts/requeue_needs_retry.ts
/**
 * Re-queues jobs marked "Needs Retry" into the pipeline before scrape/evaluate.
 * Implements CR-011, FR-035.
 */
import Database from 'better-sqlite3';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = path.join(__dirname, '..');
const DB_PATH = path.join(PROJECT_ROOT, 'data/jobagent.sqlite');
const JOBS_DIR = path.join(PROJECT_ROOT, 'jobs');
const SUBMISSIONS_DIR = path.join(PROJECT_ROOT, 'data/submissions');
const MAX_AUTO_RETRIES = 3;

function resolveSubmissionFolder(company: string): string | null {
  const slug = company.toLowerCase().replace(/[^a-z0-9]+/g, '_').replace(/^_+|_+$/g, '');
  const direct = path.join(SUBMISSIONS_DIR, slug);
  if (fs.existsSync(direct)) return direct;

  if (!fs.existsSync(SUBMISSIONS_DIR)) return null;
  const stripped = company.toLowerCase().replace(/[^a-z0-9]/g, '');
  for (const name of fs.readdirSync(SUBMISSIONS_DIR)) {
    const full = path.join(SUBMISSIONS_DIR, name);
    if (!fs.statSync(full).isDirectory()) continue;
    if (name.toLowerCase().replace(/[^a-z0-9]/g, '') === stripped) return full;
  }
  return null;
}

async function run() {
  const db = new Database(DB_PATH);

  db.prepare(`UPDATE jobs SET status = 'Needs Retry' WHERE status = 'Failed'`).run();

  const jobs = db.prepare(`
    SELECT id, company, title, url, COALESCE(retry_count, 0) AS retry_count
    FROM jobs WHERE status = 'Needs Retry'
  `).all() as { id: string; company: string; title: string; url: string | null; retry_count: number }[];

  if (jobs.length === 0) {
    console.log('No jobs marked Needs Retry.');
    db.close();
    return;
  }

  console.log(`Re-queuing ${jobs.length} job(s) marked Needs Retry...`);
  fs.mkdirSync(JOBS_DIR, { recursive: true });

  let toDrafted = 0;
  let toNew = 0;
  let skipped = 0;

  const setDrafted = db.prepare(`UPDATE jobs SET status = 'Drafted' WHERE id = ?`);
  const setNew = db.prepare(`UPDATE jobs SET status = 'New' WHERE id = ?`);

  for (const job of jobs) {
    if (job.retry_count >= MAX_AUTO_RETRIES) {
      console.log(`  -> Skip ${job.company} — max auto-retries (${MAX_AUTO_RETRIES}) reached.`);
      skipped++;
      continue;
    }

    const companyFilename = job.company.replace(/[^a-z0-9]+/gi, '_').trim();
    const stagingPath = path.join(JOBS_DIR, `${companyFilename}_${job.id.slice(0, 8)}.txt`);
    const folder = resolveSubmissionFolder(job.company);
    const originalJd = folder ? path.join(folder, 'Original_JD.txt') : null;

    if (originalJd && fs.existsSync(originalJd)) {
      let content = fs.readFileSync(originalJd, 'utf-8');
      if (job.url && !content.trimStart().startsWith('URL:')) {
        content = `URL: ${job.url}\n\n${content}`;
      }
      fs.writeFileSync(stagingPath, content, 'utf-8');
      setDrafted.run(job.id);
      console.log(`  -> ${job.company}: restored JD → Drafted (${path.basename(stagingPath)})`);
      toDrafted++;
    } else if (job.url && !job.url.startsWith('local://')) {
      setNew.run(job.id);
      console.log(`  -> ${job.company}: will re-scrape → New`);
      toNew++;
    } else {
      console.log(`  -> Skip ${job.company} — no saved JD or URL to requeue.`);
      skipped++;
    }
  }

  console.log(`[REQUEUE_DONE] ${toDrafted} → Drafted, ${toNew} → New, ${skipped} skipped.`);
  db.close();
}

run().catch((e) => {
  console.error(e);
  process.exit(1);
});
